daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -902,6 +903,105 @@ object SBuiltin {
     }
   }
 
+  // This function translates well-type values
+  private def translateValue(machine: Machine, value: V[V.ContractId]): CtrlValue = {
+    def go(value0: V[V.ContractId]): SValue =
+      value0 match {
+        case V.ValueList(vs) => SList(vs.map[SValue](go))
+        case V.ValueContractId(coid) => SContractId(coid)
+        case V.ValueInt64(x) => SInt64(x)
+        case V.ValueNumeric(x) => SNumeric(x)
+        case V.ValueText(t) => SText(t)
+        case V.ValueTimestamp(t) => STimestamp(t)
+        case V.ValueParty(p) => SParty(p)
+        case V.ValueBool(b) => SBool(b)
+        case V.ValueDate(x) => SDate(x)
+        case V.ValueUnit => SUnit
+        case V.ValueRecord(Some(id), fs) =>
+          val fields = Name.Array.ofDim(fs.length)
+          val values = new util.ArrayList[SValue](fields.length)
+          fs.foreach {
+            case (optk, v) =>
+              optk match {
+                case None =>
+                  throw crash("SValue.fromValue: record missing field name")
+                case Some(k) =>
+                  fields(values.size) = k
+                  val _ = values.add(go(v))
+              }
+          }
+          SRecord(id, fields, values)
+        case V.ValueRecord(None, _) =>
+          crash("SValue.fromValue: record missing identifier")
+        case V.ValueStruct(fs) =>
+          val fields = Name.Array.ofDim(fs.length)
+          val values = new util.ArrayList[SValue](fields.length)
+          fs.foreach {
+            case (k, v) =>
+              fields(values.size) = k
+              val _ = values.add(go(v))
+          }
+          SStruct(fields, values)
+        case V.ValueVariant(None, _variant @ _, _value @ _) =>
+          crash("SValue.fromValue: variant without identifier")
+        case V.ValueEnum(None, constructor @ _) =>
```

Comment:
Is the "positive" `ValueEnum(Some(id), ...)` missing?

Edit: Ah no, it's further down. Maybe colocate the two cases?